server/services/ai.js

```diff
@@ -28,6 +53,179 @@ module.exports = class AIService {
         player.credits = Math.max(0, player.credits);
     }
 
+    async _setupAi(game, player) {
+        player.researchingNext = 'random'; // Set up the AI for random research.
+        
+        // Make sure all stars are marked as not ignored - This is so the AI can bulk upgrade them.
+        const playerStars = this.starService.listStarsOwnedByPlayer(game.galaxy.stars, player._id);
+
+        for (let star of playerStars) {
+            this.starService.resetIgnoreBulkUpgradeStatuses(star);
+        }
+
+        // Clear out any carriers that have looped waypoints.
+        this.carrierService.clearPlayerCarrierWaypointsLooped(game, player);
+
+        const starGraph = this._computeStarGraph(game, player, playerStars);
+
+        // Star systems with computed score based on distance to enemy
+        const starScores = this._computeStarScores(game, player, playerStars, starGraph);
+
+        // Graph of carrier movements for logistics
+        const logisticsGraph = this._createLogisticsGraph(starGraph, starScores);
+        
+        const carrierLoops = this._computeCarrierLoopsFromGraph(logisticsGraph, playerStars);
+
+        const logisticsOrders = this._createCarrierOrders(carrierLoops);
+
+        player.scheduledOrders = logisticsOrders;
+    }
+
+    _computeStarGraph(game, player, playerStars) {
```

Comment:
This is where we can really do with spatial hash grids, this function will probably be very slow to execute! Leave it to me, I'll work on this one in the future.

Reply:
It should be still acceptable for normally-sized empires, I hope, and the calculation is only done on the first tick, so I think it's ok for now. I have changed the calculations to use the squared distance though, that should be slightly more efficient.